### PR TITLE
 `Any/Semi/Anti` join support mixed join conditions

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Support more join strictnesses (`LEFT/RIGHT SEMI/ANTI/ANY JOIN`) with inequal conditions which involve columns from both left and right table. e.g. t1.y < t2.y (see setting `allow_experimental_join_condition`)

Base on #60920

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Run these jobs only (required builds will be added automatically):
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with aarch64
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Deny these jobs:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 2e723fd5360649dd48aa6c3e154436bbb5730105 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64281/2e723fd5360649dd48aa6c3e154436bbb5730105/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by lgbo-ustc** [comment]
There is a cross join failure, https://s3.amazonaws.com/clickhouse-test-reports/64281/6917c8c9d79362bbcbac000bf335dc54c16cf637/ast_fuzzer__debug_/run.log

Fix it in https://github.com/ClickHouse/ClickHouse/pull/66413

**Comment by lgbo-ustc** [comment]
@vdimir  could you help to review this PR?

**Comment by zhanglistar** [comment]
@alexey-milovidov @vdimir Do you guys have time to review this PR please? This the last part feature to run tpc-ds complete. It's crutial for us and community to run tpc-ds.

**Comment by vdimir** [comment]
Sorry, it slipped my mind, I'll review it today

File: src/Interpreters/joinDispatch.h [hunk 1]
```diff
@@ -12,38 +12,41 @@
...
-template <> struct MapGetter<JoinKind::Inner, JoinStrictness::Anti> { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
-template <> struct MapGetter<JoinKind::Right, JoinStrictness::Anti> { using Map = HashJoin::MapsAll; static constexpr bool flagged = true; };
-template <> struct MapGetter<JoinKind::Full, JoinStrictness::Anti>  { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
+template <> struct MapGetter<JoinKind::Left, JoinStrictness::Anti, false>  { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
+template <> struct MapGetter<JoinKind::Left, JoinStrictness::Anti, true>  { using Map = HashJoin::MapsAll; static constexpr bool flagged = false; };
+template <bool prefer_use_maps_all> struct MapGetter<JoinKind::Inner, JoinStrictness::Anti, prefer_use_maps_all> { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
+template <bool prefer_use_maps_all> struct MapGetter<JoinKind::Right, JoinStrictness::Anti, prefer_use_maps_all> { using Map = HashJoin::MapsAll; static constexpr bool flagged = true; };
```
Comment (vdimir) [inline src/Interpreters/joinDispatch.h:45]: It's a bit unclear that `prefer_use_maps_all` affects the left anti join, while for inner anti join, it is still `MapsOne`. Could we avoid introducing `prefer_use_maps_all` in `MapGetter`, and instead override `MapGetter::Map` in the locations where we use it based on the kind and strictness? However, if this makes the code less readable, let's keep it as is. Just add a comment about `prefer_use_maps_all`, especially in cases where `Map = HashJoin::MapsOne` is used.

File: src/Interpreters/joinDispatch.h [hunk 2]
```diff
@@ -12,38 +12,41 @@
...
 
 /// Only SEMI LEFT and SEMI RIGHT are valid. INNER and FULL are here for templates instantiation.
-template <> struct MapGetter<JoinKind::Left, JoinStrictness::Anti>  { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
-template <> struct MapGetter<JoinKind::Inner, JoinStrictness::Anti> { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
-template <> struct MapGetter<JoinKind::Right, JoinStrictness::Anti> { using Map = HashJoin::MapsAll; static constexpr bool flagged = true; };
-template <> struct MapGetter<JoinKind::Full, JoinStrictness::Anti>  { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
+template <> struct MapGetter<JoinKind::Left, JoinStrictness::Anti, false>  { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
+template <> struct MapGetter<JoinKind::Left, JoinStrictness::Anti, true>  { using Map = HashJoin::MapsAll; static constexpr bool flagged = false; };
+template <bool prefer_use_maps_all> struct MapGetter<JoinKind::Inner, JoinStrictness::Anti, prefer_use_maps_all> { using Map = HashJoin::MapsOne; static constexpr bool flagged = false; };
```
Comment (vdimir) [inline src/Interpreters/joinDispatch.h:44]: Why we can use `MapsOne` for INNER ? Shouldn't we still have all combinations to check which are matched and which is not?

File: src/Interpreters/joinDispatch.h (hunk 1 above)
Comment (lgbo-ustc) [reply src/Interpreters/joinDispatch.h:45]: Need to introduce new strictnesses. These strictnesses are not visible for users, and they will be assigned in analyzer

File: src/Interpreters/joinDispatch.h (hunk 2 above)
Comment (lgbo-ustc) [reply src/Interpreters/joinDispatch.h:44]: I miss `INNER ANY` here， it should be supported here. `INNER SEMI` and `INNER ANTI` are invalid as comments say， so I left them all `MapsOne`

**Review by vdimir — APPROVED** [review]

**Pull request merged**